### PR TITLE
Support row filter buffering

### LINKED ISSUE #270: Implement filter_buffer
Aim:
Extend the functionality of "row_filter" to filter "n" samples that do not represent normal running condition. 

Description:
The "row_filter" filters the data that not match a certain condition. Once this condition is met again, the "filter_buffer" filters the next "n" number of samples.

Proposed placeholder in config file:

In bold, you can find the suggested placeholder for the "filter_buffer", added as a parameter in the dataset block:

- name: cooling
    dataset:
      tags:
      - GRA-SE  -23-1290.PV
      - GRA-TE  -23-0735A.PV
      - GRA-TE  -23-0735B.PV
      - GRA-TE  -23-0735C.PV
      - GRA-TE  -23-0735D.PV
      - GRA-TE  -23-0737A.PV
      - GRA-TE  -23-0737B.PV
      - GRA-TE  -23-0737C.PV
      - GRA-TE  -23-0737D.PV
      - GRA-NRA -23-0413.PV
      train_start_date: 2015-07-02T00:00:00+01:00
      train_end_date: 2015-09-15T00:00:00+01:00
      row_filter="`GRA-NRA -23-0413.PV` > 8500"
      **filter_buffer=12**
    model:
      sklearn.pipeline.Pipeline:
        steps:
        - sklearn.pipeline.FeatureUnion:
          - sklearn.preprocessing.data.MinMaxScaler
        - gordo_components.model.models.KerasAutoEncoder:
            kind: feedforward_hourglass

### PR DESCRIPTION
**Pull request by milesgranger**
Will close #270 

Problem :chart_with_downwards_trend: 
---
We support a superb `row_filter` param to our `TimeSeriesDataset`, however, there is an additional request that areas fore and aft of such filtering is also filtered out to account for "ramp up" and "ramp down" trends in the time series.

Solution :chart_with_upwards_trend:  
---
Add another parameter to `TimeSeriesDataset` which is passed down into `pandas_filter_rows` which the resulting mask is 'expanded' to include filtering out indexes within the buffer zone.

### CODE REVIEW

File: gordo_components/dataset/datasets.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def __init__(
         resolution: str = "10T",
         row_filter: str = "",
         aggregation_methods: Union[str, List[str], Callable] = "mean",
+        row_filter_buffer_size: int = 0,
```
Comment (flikka) [inline gordo_components/dataset/datasets.py:30]: Some sweet docstring missing here?

File: gordo_components/dataset/filter_rows.py [hunk 2]
```diff
@@ -41,7 +43,44 @@ def visit_Str(self, node):
...
+    None
+    """
+    idxs, *_rows = np.where(mask == False)
+    for idx in idxs:
+        mask.values[
+            range(max((0, idx - buffer_size)), min((len(mask), idx + buffer_size + 1)))
+        ] = False
+
+
+def pandas_filter_rows(df, filter_str: str, buffer_size: int = 0):
```
Comment (flikka) [inline gordo_components/dataset/filter_rows.py:83]: Some docs missing here?

**Review by flikka — APPROVED** [review]
Elegant and wonderful. Some docs and we're good to go!

**Comment by milesgranger** [comment]
@ruimaximo This is what you were after, starting with `0.37.0-rc2` you can add `row_filter_buffer_size` to the `dataset` key in the config file. :+1:
